File: apps/websocket-skyblock-csharp/src/hypixel/routes/AuctionsRoute/AuctionsRoute.cs
using System.Text.Json.Nodes;
using Newtonsoft.Json;
using ZstdSharp.Unsafe;

public class AuctionsRoute
{
    public bool success;
    public int page;
    public int totalPages;
    public int totalAuctions;
    public double lastUpdated;
    public AuctionsRouteProduct[] auctions = default!;

    /// <summary>
    /// Note: The optional cache is meant to be used on all incoming items, hopefully making subsequent actions easier.
    /// </summary>
    /// <param name="CachedItems"></param>
    /// <returns></returns>
    public static async Task<List<AuctionsRouteProduct>?> GetRoute()
    {
        // there will be 2 fetch types:
        // > Default Fetch: fetches all pages, 5minute cooldown
        // > Max Fetch: fetches latest 5 pages, 1minute cooldown
        // 
        // Max Fetch will occur during the following:
        // - very start of the application launch
        // - when there is a new/modified item in the feed
        // 
        // otherwise, use the Default Fetch
        List<AuctionsRouteProduct>? auctions;

        if (MongoBot.AuctionBuysRecentlyUpdated)
        {
            auctions = await BatchFetch(10);
        }
        else
        {
            auctions = await BatchFetch(100);
            MongoBot.AuctionBuysRecentlyUpdated = true;
        }

        if (auctions == null)
            return null;

        return auctions;
    }

    static async Task<List<AuctionsRouteProduct>?> BatchFetch(int pages)
    {
        List<AuctionsRouteProduct> products = new List<AuctionsRouteProduct>();
        int i = 0;

        try
        {
            AuctionsRoute? firstResults = await GetResult(0);

            if (firstResults == null)
                return null;

            // NOTE: im doing it this way instead of an array of tasks, because for some reason, an async function within a task has unexpected results when awaiting (skips it + the rest of the code)
            products.AddRange(firstResults.auctions);
            int totalPages = Math.Min(firstResults.totalAuctions / 1000, pages);
            Console.WriteLine(firstResults.totalAuctions);
            Action?[] tasks = new Action[totalPages - 1];
            Task dummyTask = new Task(() => { }); // used to mimic all tasks being done
            List<AuctionsRouteProduct>[] fetchedProducts = new List<AuctionsRouteProduct>[totalPages - 1];

            for (i = 1; i < totalPages; i++)
            {
                int page = i;
                Action task = async () =>
                {
                    Utility.Log(Enums.LogLevel.NONE, $"auction page {page}", false, false);
                    AuctionsRoute? auctionsRoute = await GetResult(page);

                    if (auctionsRoute == null)
                        return;

                    products.AddRange(auctionsRoute.auctions);
                    fetchedProducts[page - 1] = auctionsRoute.auctions.ToList();
                    tasks[page - 1] = null;

                    if (tasks.All(e => e == null))
                        dummyTask.Start();
                };

                tasks[page - 1] = task;
            }

            foreach (Action? task in tasks)
                task?.Invoke();

            Task.WaitAll(dummyTask);

            return products;
        }

        catch (Exception e)
        {
            Utility.Log(Enums.LogLevel.ERROR, $"Unexpected error on page {i}: {e}");
            return null;
        }
    }

    static async Task<AuctionsRoute?> GetResult(int page)
    {
        string json;
        int maxRetry = 3;
        int currentRetry = 0;

        try
        {
            json = await Program.Client.GetStringAsync($"{Settings.HYPIXEL_API_BASE_URL}/skyblock/auctions?page={page}");
        }

        catch (Exception)
        {
            currentRetry++;
            Utility.Log(Enums.LogLevel.WARN, $"Fetch failed! ({currentRetry})");
            Thread.Sleep(500);

            if (currentRetry >= maxRetry)
                return null;

            json = await Program.Client.GetStringAsync($"{Settings.HYPIXEL_API_BASE_URL}/skyblock/auctions?page={page}");
        }

        JsonNode? result = JsonArray.Parse(json);

        if (result == null)
        {
            Utility.Log(Enums.LogLevel.ERROR, "Unexpected null when the result should not be null.");
            return null;
        }

        if (result["success"] == null)
        {
            Utility.Log(Enums.LogLevel.ERROR, "Unexpected null when the result should always contains a 'success' field.");
            return null;
        }

        if (result["success"]!.Equals(false))
        {
            Utility.Log(Enums.LogLevel.WARN, "Success = false, returning...");
            return null;
        }

        AuctionsRoute? auctionsRoute = JsonConvert.DeserializeObject<AuctionsRoute>(json);

        if (auctionsRoute == null)
        {
            Utility.Log(Enums.LogLevel.ERROR, "Unexpected null while converting response to C# class.");
            return null;
        }

        return auctionsRoute;
    }
}

File: apps/discord-bot/app/src/main/java/discordbot/hypixel/routes/AuctionsRoute/AuctionsRoute.java
package discordbot.hypixel.routes.AuctionsRoute;

import com.google.gson.Gson;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse.BodyHandlers;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.HashMap;
import discordbot.BotSettings;
import discordbot.Main;
import discordbot.common.Enums.LogLevel;
import discordbot.common.Enums.Settings;

public class AuctionsRoute {
    public boolean success;
    public int page;
    public int totalPages;
    public int totalAuctions;
    public double lastUpdated;
    public AuctionsRouteProduct[] auctions;

    public static ArrayList<AuctionsRouteProduct> GetRoute() {
        return GetRoute(null, 100);
    }

    public static ArrayList<AuctionsRouteProduct> GetRoute(HashMap<String, Double> CachedItems, int pages) {
        ArrayList<AuctionsRouteProduct> auctions = new ArrayList<AuctionsRouteProduct>();
        int i = 0;

        try {

            for (i = 0; i < pages; i++) {
                // Main.Utility.LogPerformance("Total Start", false);

                String json = new String("");
                int maxRetry = 3;
                int currentRetry = 0;

                do {
                    try {
                        URI uri = new URI(MessageFormat.format("{0}/skyblock/auctions?page={1}", BotSettings.Get(Settings.HYPIXEL_API_BASE_URL), i));
                        HttpRequest request = HttpRequest.newBuilder(uri).header("API-Key", BotSettings.Get(Settings.HYPIXEL_BOT_KEY)).build();
                        json = Main.Client.send(request, BodyHandlers.ofString()).body();
                        Main.Utility.Log(LogLevel.NONE, json);
                        AuctionsRouteProduct product = Main.Gson.fromJson(json, AuctionsRouteProduct.class);
                        // System.out.println(product.);
                        currentRetry = maxRetry + 1; // exit loop
                    } catch (Exception e) {
                        Thread.sleep(100);
                        currentRetry += 1;
                        Main.Utility.Log(LogLevel.WARN, MessageFormat.format("Fetch failed! ({0})", e));

                        if (currentRetry >= maxRetry)
                            throw new Exception(MessageFormat.format("Reached max retry limit! Throwing... {0}", e));
                    }
                } while (currentRetry <= maxRetry);

            }

        } catch (Exception err) {

        }

        return auctions;
    }
}

//                 async Task setJson() => json = await Program.Client.GetStringAsync($"{Settings.HYPIXEL_API_BASE_URL}/skyblock/auctions?page={i}");

//                 try
//                 {
//                     await setJson();
//                 }

//                 catch (Exception)
//                 {
//                     currentRetry++;
//                     Thread.Sleep(100);

//                     Program.Utility.Log(Enums.LogLevel.WARN, $"Fetch failed! ({currentRetry})");
//                     if (currentRetry >= maxRetry)
//                         throw;

//                     await setJson();
//                 }

//                 JsonNode? result = JsonArray.Parse(json);

//                 if (result == null)
//                 {
//                     Program.Utility.Log(Enums.LogLevel.ERROR, "Unexpected null when the result should not be null.");
//                     return null;
//                 }

//                 if (result["success"] == null)
//                 {
//                     Program.Utility.Log(Enums.LogLevel.ERROR, "Unexpected null when the result should always contains a 'success' field.");
//                     return null;
//                 }

//                 if (result["success"]!.Equals(false))
//                 {
//                     Program.Utility.Log(Enums.LogLevel.WARN, "Success = false, returning...");
//                     return null;
//                 }

//                 AuctionsRoute? auctionsRoute = JsonConvert.DeserializeObject<AuctionsRoute>(json);

//                 if (auctionsRoute == null)
//                 {
//                     Program.Utility.Log(Enums.LogLevel.ERROR, "Unexpected null while converting response to C# class.");
//                     return null;
//                 }

//                 foreach (AuctionsRouteProduct product in auctionsRoute.auctions)
//                 {
//                     if (CachedItems != null)
//                         if (!CachedItems.ContainsKey(product.uuid))
//                             CachedItems.Add(product.uuid, default);
//                         else
//                             continue;

//                     auctions.Add(product);
//                 }

//                 Program.Utility.Log(Enums.LogLevel.NONE, $"Page {i + 1} of auctions fetched.", true, false);

//                 if (auctionsRoute.auctions.Length < 1000)
//                 {
//                     Program.Utility.Log(Enums.LogLevel.NONE, $"Found end of length at page {i + 1}. Closing...");
//                     break;
//                 }

//                 await Task.Delay(10); // make sure we dont request too fast
//             }

//             return auctions;
//         }

//         catch (Exception e)
//         {
//             Program.Utility.Log(Enums.LogLevel.ERROR, $"Unexpected error on page {i}: {e}");
//             return null;
//         }
//     }
// }